Fixed differently upstream by commit bdcd06c4cc1971d763e528b8cb1d0f16fcc5ecf4

--- boost/boost/concept/detail/general.hpp.orig	2015-07-31 14:03:08.184447736 +0200
+++ boost/boost/concept/detail/general.hpp	2015-07-31 14:03:10.048447591 +0200
@@ -67,7 +67,7 @@
 
 // Version check from https://svn.boost.org/trac/boost/changeset/82886
 // (boost/static_assert.hpp)
-#if defined(__GNUC__) && ((__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 7))) 
+#if defined(__GNUC__) && (((__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 7))) || defined(__clang__))
 #define BOOST_CONCEPT_UNUSED_TYPEDEF __attribute__((unused))
 #else
 #define BOOST_CONCEPT_UNUSED_TYPEDEF /**/
